File: leetcode/cpp/973. 最接近原点的 K 个点.cpp
class Solution {
public:
    // 交换函数
    void swap(std::vector<std::vector<int>>& vec, int a, int b) {
        auto tmp = vec[a];
        vec[a] = vec[b];
        vec[b] = tmp;
    }

    // 获取点到原点的距离平方
    int getDistance(std::vector<int>& point) {
        return point[0]*point[0] + point[1]*point[1];
    }

    // 堆化函数，小顶堆
    void heapify(std::vector<std::vector<int>>& vec, int n, int i) {
        int largest = i;
        int l = 2 * i + 1;
        int r = 2 * i + 2;
        if (l < n && getDistance(vec[l]) < getDistance(vec[largest])) {
            largest = l;
        }
        if (r < n && getDistance(vec[r]) < getDistance(vec[largest])) {
            largest = r;
        }
        if (largest != i) {
            swap(vec, i, largest);
            heapify(vec, n, largest);
        }
    }

    vector<vector<int>> kClosest(vector<vector<int>>& points, int k) {
        // 特殊情况
        if (k <= 0 || k > points.size()) {
            return std::vector<std::vector<int>>{};
        }

        // 建立小顶堆
        int n = points.size();
        for (int i = n/2-1; i >= 0; i--) {
            heapify(points, n, i);
        }

        // 获取topk，移动前k小到末尾，末尾最小
        for (int i = n - 1; i >= n - k; i--) {
            swap(points, 0, i);
            heapify(points, i, 0);
        }

        // 获取结果
        std::vector<std::vector<int>> ret;
        for (int i = 0; i < k; i++) {
            ret.push_back(points[n - 1 - i]);
        }
        return ret;
    }

	// ===== 另一种方式（快排）=====
    // 快排递归函数
    void quickSort(std::vector<std::vector<int>>& vec, int l, int h) {
        if (l >= h) {
            return;
        }
        int m = partition(vec, l, h);
        quickSort(vec, l, m - 1);
        quickSort(vec, m + 1, h);
        return;
    }

    vector<vector<int>> kClosest(vector<vector<int>>& points, int k) {
        // 特殊情况
        if (k <= 0 || k > points.size()) {
            return std::vector<std::vector<int>>{};
        }

        // 二分得到topk，索引为k-1，找到后排序
        int l = 0, h = points.size() - 1;
        while (l <= h) {
            int m = partition(points, l, h);
            if (m == k - 1) {
                quickSort(points, 0, k - 1);
                break;
            } else if (m > k - 1) {
                h = m - 1;
            } else if (m < k - 1) {
                l = m + 1;
            }
        }

        // 获取结果
        std::vector<std::vector<int>> ret;
        for (int i = 0; i < k; i++) {
            ret.push_back(points[i]);
        }
        return ret;
    }
};

File: leetcode/python/973. 最接近原点的 K 个点.py
#!/bin/bash python
"""
973. 最接近原点的 K 个点

思路：
堆排序，快排
"""

class Solution(object):
    def heapify(self, points, distances, n, i):
        """
        堆化函数（维护小顶堆，小元素上浮）
        """
        smallest = i
        l = 2 * i + 1
        r = 2 * i + 2
        if l < n and distances[l] < distances[smallest]:
            smallest = l
        if r < n and distances[r] < distances[smallest]:
            smallest = r
        if smallest != i:
            points[i], points[smallest] = points[smallest], points[i]
            distances[i], distances[smallest] = distances[smallest], distances[i]
            self.heapify(points, distances, n, smallest)

    def kClosest(self, points, k):
        """
        :type points: List[List[int]]
        :type k: int
        :rtype: List[List[int]]
        """
        n = len(points)
        if points == 0 or k >= n:
            return points
        
        # 获取对应的距离平方数组
        distances = [0] * n
        for i in range(n):
            distances[i] = points[i][0]**2 + points[i][1]**2
	
	# ===== 另一种方式 =====
    def partition(self, points, l, h):
        """
        快排分区函数
        """
        start = l - 1
        for i in range(l, h):
            if points[i][0]**2+points[i][1]**2 < points[h][0]**2+points[h][1]**2:
                start += 1
                points[i], points[start] = points[start], points[i]
        start += 1
        points[h], points[start] = points[start], points[h]
        return start
    
    def kClosest(self, points: List[List[int]], k: int) -> List[List[int]]:
        """
        使用快排的思路
        """
        if points == [] or k > len(points):
            return points
        
        # 找到第k大的元素，索引为k-1
        l, h = 0, len(points) - 1
        while l < h:
            m = self.partition(points, l, h)
            if m == k - 1:
                break
            elif m > k - 1:
                h = m - 1
            elif m < k - 1:
                l = m + 1
        
        # 返回前k小的元素
        return points[:k]